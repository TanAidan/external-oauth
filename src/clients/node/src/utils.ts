import Types from './types'

/**
 * Helper to build a new URL from different params
 */
export function toURL(origin: string, baseURL: string, endpoint: string, queryString?: Types.RequestQueryString): URL {
  const removeLeadingSlash = (text: string) => {
    return text.replace(/^\//, '')
  }

  const removeTrailingSlash = (text: string) => {
    return text.replace(/\/$/, '')
  }

  const urlParts: string[] = []
  urlParts.push(removeTrailingSlash(origin))
  urlParts.push(removeLeadingSlash(removeTrailingSlash(baseURL)))
  urlParts.push(removeLeadingSlash(endpoint))

  const url = new URL(urlParts.join('/'))

  if (queryString) {
    Object.keys(queryString).forEach(key => url.searchParams.append(key, String(queryString[key])))
  }

  return url
}

/**
 * Helper to remove all undefined keys
 * @param obj {object}
 */
export function cleanHeaders(obj: Record<string, any>) {
  return Object.keys(obj).reduce((acc, key: string) => {
    if (obj[key] !== undefined) {
      acc[key] = obj[key]
    }
    return acc
  }, {} as Record<string, any>)
}
