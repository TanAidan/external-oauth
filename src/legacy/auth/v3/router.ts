import { Router } from 'express'
import { connectContext, callbackContext } from './context'
import { connectConfig, callbackConfig } from './config'
import { session, destroySession, destroySessionOnError } from './session'
import { authenticate, fetchAuthDetails } from './strategy'
import { errorHandler } from './error-handler'
import { authSuccess } from './success'
import { connectAuthId, callbackAuthId } from './auth-id'
import { revoke } from './revoke'
import { connectSetupId } from './setup-id'
import { compose } from 'compose-middleware'
import { configureAuthDetailsRequest } from '../configure-request'
import { authDetailsResponse } from '../auth-details'
import { connectBuid } from '../../middlewares/set-identifiers'
import { setupId } from '../../functions/router'

export default () => {
  const authenticateAndRespond = compose(authenticate, destroySession, authSuccess)

  const router = Router()

  router.use(session())

  router.get('/callback', callbackContext, callbackConfig, callbackAuthId, authenticateAndRespond)

  router.get(
    '/:buid',
    connectBuid,
    connectContext,
    setupId,
    connectSetupId,
    connectConfig,
    connectAuthId,
    authenticateAndRespond
  )

  router.delete('/:buid/revoke/:authId', revoke)

  router.use(destroySessionOnError)
  router.use(errorHandler)

  return router
}

export const authRouter = () => {
  const router = Router()

  router.get('/:buid/auth/:authId', configureAuthDetailsRequest, setupId, fetchAuthDetails, authDetailsResponse)

  return router
}
