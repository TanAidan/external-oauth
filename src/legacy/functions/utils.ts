// Is the given object empty e.g. {}?
export function isEmpty(obj: Object) {
  return Object.getOwnPropertyNames(obj).length === 0
}
