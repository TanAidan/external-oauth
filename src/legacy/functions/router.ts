import { Router, NextFunction } from 'express'

import errorHandler from '../../legacy/errorHandler'
import { proxyCorsMiddleware } from '../proxy/cors'
import { fetchAuthDetails } from '../auth/v3/strategy'
import { middleware as proxyHandler } from './lambda-request'
import { PROXY_PREFIX, setProxyFunction } from '../middlewares/v4/intent-info'
import Integration from './integration'
import { getSetupDetails } from '../auth/clients/integrations'
import { PizzlyError } from '../../lib/error-handling'

const BUID = 'bearerUuid'

function functionsRouter() {
  const functionsRouter = Router({ mergeParams: true })

  functionsRouter.all(`${PROXY_PREFIX}/*`, setupId, fetchAuthDetails, setProxyFunction, proxyHandler())
  functionsRouter.use((_req, res) => res.status(404).send())
  functionsRouter.use(errorHandler)

  return functionsRouter
}

function proxyRouter() {
  const proxyRouter = Router({ mergeParams: true })

  proxyRouter.all(`/*`, fetchAuthDetails, setProxyFunction, proxyHandler)
  return proxyRouter
}

function setProxyCallType(req: any, _res, next: NextFunction) {
  const apiKey = (req.headers && req.headers.authorization) || req.query.apiKey
  if (apiKey) {
    req.isBackend = true
  } else {
    req.isBackend = false
  }

  next()
}

export default () => {
  const router = Router()
  router.use(`/backend/:${BUID}`, buid, authId, isBackendRequest, functionsRouter())
  router.use(`/:${BUID}`, buid, authId, functionsRouter())
  router.use(errorHandler)

  return router
}

export function proxyFunction() {
  const router = Router()
  router.use(`/:${BUID}`, proxyCorsMiddleware, buid, setupId, authId, setProxyCallType, proxyRouter())
  router.use(errorHandler)
  return router
}

function authId(req, _res, next) {
  req.authId = req.headers['bearer-auth-id'] || req.headers['pizzly-auth-id'] || req.query.authId
  next()
}

function buid(req, _res, next) {
  req.buid = req.params[BUID]
  req.integration = new Integration(req.buid)
  next()
}

export async function setupId(req, _res, next) {
  const configuration = await getSetupDetails({ buid: req.buid, store: req.store, setupId: req.query.setupId })

  if (!configuration) {
    return next(new PizzlyError('unknown_configuration'))
  }

  req.setupId = configuration.setupId
  req.configuration = configuration

  next()
}

function isBackendRequest(req, _res, next) {
  req.isBackend = true
  next()
}
