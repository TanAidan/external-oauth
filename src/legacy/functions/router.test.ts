const check = jest.fn()
const PAYLOAD = '{"data":{"myData":"ok","status":"200","data":{"data":{"myData":"ok"}}}}'
const handler = jest.fn((req, res, next) => {
  check({ ...req })
  req.bearerResponse = {
    Payload: PAYLOAD,
    StatusCode: 200
  }
  next()
})

const fetchAuthDetails = (req, _res, next) => {
  req.auth = {
    accessToken: 'test-access-token'
  }

  next()
}

jest.mock('../auth/v3/strategy', () => ({
  fetchAuthDetails,
  fetchAuthDetailsForFunctions: fetchAuthDetails
}))

import request from 'supertest'

import { baseApp } from '../../../tests/utils'
import router, { proxyFunction } from './router'

const app = baseApp()
app.use('/whatever', router())
app.use('/proxy', proxyFunction())

describe('proxyFunction', () => {
  beforeEach(async () => {
    check.mockClear()
    handler.mockClear()
  })

  describe('authenticate with api key and authId', () => {
    const setup = async () => {
      await request(app)
        .get('/proxy/github/user/repos')
        .set('Authorization', 'sk_my-api-key')
        .set('Bearer-Auth-Id', 'auth id')
        .expect(200, { data: { myData: 'ok' } })
    }

    it('sets req with expected variables', async () => {
      await setup()
      expect(handler).toHaveBeenCalledTimes(1)
      const req = check.mock.calls[0][0]
      expect(req.isBackend).toBe(true)
      expect(req.headers.authorization).toBe('sk_my-api-key')
      expect(req.headers['bearer-auth-id']).toBe('auth id')
    })
  })

  describe('authenticate with bearer-publishable-key', () => {
    const userCorrelationId = 'test-corr-id'
    const setup = async () => {
      await request(app)
        .get('/proxy/github/user/repos')
        .set('Bearer-Publishable-Key', 'pk_my-api-key')
        .set('Bearer-Auth-Id', 'auth id')
        .set('Bearer-Correlation-Id', userCorrelationId)
        .expect(200, { data: { myData: 'ok' } })
    }
    it('sets req with expected variables', async () => {
      await setup()
      expect(handler).toHaveBeenCalledTimes(1)
      const req = check.mock.calls[0][0]
      expect(req.isBackend).toBe(false)
      expect(req.headers.authorization).toBeUndefined()
      expect(req.headers['bearer-publishable-key']).toBe('pk_my-api-key')
      expect(req.headers['bearer-auth-id']).toBe('auth id')
    })
  })
})
