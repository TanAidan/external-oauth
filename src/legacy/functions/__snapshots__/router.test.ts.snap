// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proxyFunction authenticate with api key and authId sets req with expected variables 1`] = `"arn:aws:lambda:eu-west-3:1:function:gateway-test-service"`;

exports[`proxyFunction authenticate with bearer-publishable-key sets req with expected variables 1`] = `"arn:aws:lambda:eu-west-3:1:function:gateway-test-service"`;
